app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -26,40 +31,72 @@ def each_batch
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
+    filter = "eventTimestamp ge #{most_recent_time}"
+    fields = 'authorization,correlationId,description,eventDataId,eventName,eventTimestamp,operationName,resourceGroupName,resourceProviderName,resourceId,resourceType'
+
+    events = connection.list(:filter => filter, :select => fields, :all => true)
+
+    if events.present?
+      existing_records = EventStream.select(:ems_ref).where(:source => 'AZURE', :ems_ref => events.map(&:event_data_id)).map(&:ems_ref)
+      events = events.reject{ |e| existing_records.include?(e.event_data_id) } if existing_records.present?
+    end
 
-    # HACK: the Azure Insights API does not support the 'gt' (greater than relational operator)
-    # therefore we have to poll from 1 millisecond past the timestamp of the last event to avoid
-    # gathering the same event more than once.
-    @since = one_ms_from_last_timestamp(events) unless events.empty?
     events
   end
 
+  # When the appliance first starts, or is restarted, start looking for events
+  # from a fixed, recent point in the past.
+  #
   def startup_interval
-    format_timestamp(1.minute.ago)
+    format_timestamp(2.minutes.ago)
   end
 
-  def one_ms_from_last_timestamp(events)
-    time = Time.at(one_ms_from_last_timestamp_as_f(events)).utc
-    format_timestamp(time)
-  end
+  # Retrieve the most recent Azure event minus 2 minutes, or the startup interval
+  # if no records are found.
+  #
+  # Go back a maximum of 1 hour if the newest record is older than that to avoid
+  # hitting the account's request limits. This should only happen in practice if
+  # the appliance has been down for a while and was restarted.
+  #
+  def most_recent_time
+    result = EventStream.select(:timestamp).where(:source => 'AZURE').order('timestamp desc').limit(1).first
```

Comment:
Ok, updated.

Not sure what you mean by filtering by ems id. We want the most recent time of the latest Azure event.

Reply:
Can you theoretically have two and more records for Azure provider? With different credentials of different departments of some organization, in the two different regions? If so, `event_streams` table will have records with different `ems_id`. You'll get the most recent timestamp from both of those emses, but it will only be valid for one of them.

Reply:
And we still have direct database access here...

Reply:
> Not sure what you mean by filtering by ems id

ok, Adam explained better [here](https://github.com/ManageIQ/manageiq-providers-azure/pull/350#discussion_r326146682).

Reply:
Updated.